http-verbs-common/src/main/scala/uk/gov/hmrc/http/HttpReadsInstances.scala

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright 2020 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.http
+
+import play.api.libs.json.{JsValue, JsError, JsResult, JsSuccess, Reads => JsonReads}
+import scala.util.{Failure, Success, Try}
+
+trait HttpReadsInstances extends HttpReadsHttpResponse with HttpReadsEither with HttpReadsOption with HttpReadsJson
+
+object HttpReadsInstances extends HttpReadsInstances
+
+import HttpReadsInstances._
+
+trait HttpReadsHttpResponse {
+  /** returns the HttpResponse as is - you will be responsible for checking any status codes. */
+  implicit val readRaw: HttpReads[HttpResponse] =
+    HttpReads.ask.map { case (_, _, response) => response }
+}
+
+trait HttpReadsEither {
+  implicit def readEitherOf[A : HttpReads]: HttpReads[Either[UpstreamErrorResponse, A]] =
+    HttpReads.ask.flatMap { case (method, url, response) =>
+      HttpErrorFunctions.handleResponseEither(method, url)(response) match {
+        case Left(err)       => HttpReads.pure(Left(err))
+        case Right(response) => HttpReads[A].map(Right.apply)
+      }
+    }
+
+  def throwOnFailure[A](reads: HttpReads[Either[UpstreamErrorResponse, A]]): HttpReads[A] =
+    reads
+      .map {
+        case Left(err)    => throw err
+        case Right(value) => value
+      }
+}
+
+trait HttpReadsOption {
+  /** An opinionated HttpReads which returns None for 404/204.
+    * If you need a None for any UpstreamErrorResponse, consider using:
+    * {{{
+    *  HttpReads[Either[UpstreamErrorResponse, A]].map(_.toOption)
+    * }}}
+    */
+  implicit def readOptionOf[A : HttpReads]: HttpReads[Option[A]] =
```

Comment:
This exists in `HttpReadsLegacyInstances` doesn't it? Should this one return a response for 2xx and None for everything else?

I actually think 204 should return Some of Unit too

Reply:
Yes, I have just left this as it was defined previously.
I agree that 204 should return `Some(())`
We use this quite a lot for `404 -> None` behaviour. We could replace this with a more intuitive definition which returns None for all non-2xx and un-parseable responses, but then we'd be unaware of any non 404 errors. As much as I think this definition is a bit random, it does seem to be useful...
Alternative suggestions welcome!